File: web/src/app/services/scouts.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Scout } from 'src/app/model/Scout';
import { Role } from '../model/Role';
import { REST, RestService } from '../web/rest.service';

export interface ScoutPayload {
  name: string;
  surname: string;
  pesel: string;
  birthDate: Date;
  address: string;
  postalCode: string;
  city: string;
  phone: string;
  troopId: number;
  rankId: number;
  instructorRankId: number;
}

@Injectable({
  providedIn: 'root',
})
export class ScoutsService {
  constructor(private rest: RestService) {}

  getScouts(): Observable<Scout[]> {
    return this.rest.resolve<Scout[]>({
      method: REST.GET,
      url: `/scouts`,
    });
  }

  addRole(scoutId: number, roleId: number): Observable<boolean> {
    return this.rest.resolve<boolean>({
      method: REST.POST,
      url: `/scouts${scoutId}/roles${roleId}`,
    });
  }

  addRoles(scoutId: number, rolesIds: number[]): Observable<boolean> {
    return this.rest.resolve<boolean>({
      method: REST.POST,
      url: `/scouts${scoutId}/roles`,
      body: {
        rolesIds,
      },
    });
  }

  getRoles(scoutId: number): Observable<Role[]> {
    return this.rest.resolve<Role[]>({
      method: REST.GET,
      url: `/scouts${scoutId}/roles`,
    });
  }

  getAllRoles(): Observable<Role[]> {
    return this.rest.resolve<Role[]>({
      method: REST.GET,
      url: `/scouts/roles`,
    });
  }

  addScout(scoutPayload: ScoutPayload): Observable<boolean> {
    return this.rest.resolve<boolean>({
      method: REST.POST,
      url: `/scouts`,
      body: scoutPayload,
    });
  }

  getScout(scoutId: number): Observable<Scout> {
    return this.rest.resolve<Scout>({
      method: REST.GET,
      url: `/scouts${scoutId}`,
    });
  }

  patchScout(scoutId: number, scoutPayload: ScoutPayload): Observable<boolean> {
    return this.rest.resolve<boolean>({
      method: REST.PATCH,
      url: `/scouts${scoutId}`,
      body: scoutPayload,
    });
  }

  deleteScout(scoutId: number): Observable<boolean> {
    return this.rest.resolve<boolean>({
      method: REST.DELETE,
      url: `/scouts${scoutId}`,
    });
  }
}

File: server/src/main/java/com/app/server/repository/scoutsService/ScoutsService.java
package com.app.server.repository.scoutsService;

import com.app.server.model.data.Scout;

import java.util.Date;
import java.util.List;

public interface ScoutsService {

    Integer add(String name,
                String surname,
                String pesel,
                Date birthDate,
                String address,
                String postalCode,
                String city,
                String phone,
                int patrolId,
                int rankId,
                int instructorRankId,
                int teamId);

    Boolean addRole(int scoutId, int roleId);

    List<Scout> getAllByTeamId(int teamId);

    List<Scout> getAllByTeamIdPatrolId(int teamId, int patrolId);

    Scout getById(int scoutId);

    Boolean update(int scoutId,
                   String name,
                   String surname,
                   String pesel,
                   Date birthDate,
                   String address,
                   String postalCode,
                   String city,
                   String phone,
                   int patrolId,
                   int rankId,
                   int instructorRankId);

    Boolean deleteById(int scoutId);

    Boolean deleteRole(int scoutId, int roleId);
}
